src/client/GaloisField.cpp

```diff
@@ -0,0 +1,329 @@
+ * that Vz=y. The output z will be placed in y.
+ *
+ * @param x   the std::vector which describe the Vandermonde matrix
+ * @param y   right-hand side of the Vandermonde system equation. will be
+ *            replaced the output in this std::vector
+ * @param len consider x and y only from 0...len-1
+ */
+void GaloisField::solveVandermondeSystem(std::vector<int> &x, std::vector<int> &y, int len) {
+    for (int i = 0; i < len - 1; i++) {
+        for (int j = len - 1; j > i; j--) {
+            y[j] = y[j] ^ mulTable[x[i]][y[j - 1]];
+        }
+    }
+    for (int i = len - 1; i >= 0; i--) {
+        for (int j = i + 1; j < len; j++) {
+            y[j] = divTable[y[j]][x[j] ^ x[j - i - 1]];
+        }
+        for (int j = i; j < len - 1; j++) {
+            y[j] = y[j] ^ y[j + 1];
+        }
+    }
+}
+
+/**
+ * A "bulk" version of the substitute.
+ * Tends to be 2X faster than the "int" substitute in a loop.
+ *
+ * @param p input polynomial
+ * @param q store the return result
+ * @param x input field
+ */
+void GaloisField::substitute(std::vector< std::vector<int8_t> > &p, std::vector<int8_t> &q, int x) {
+    int y = 1;
+    for (int i = 0; i < (int)p.size(); i++) {
+        std::vector<int8_t> pi = p[i];
+        for (int j = 0; j < (int)pi.size(); j++) {
+            int pij = pi[j] & 0x000000FF;
+            q[j] = (int8_t)(q[j] ^ mulTable[pij][y]);
+        }
+        y = mulTable[x][y];
+    }
+}
+
+void GaloisField::substitute(std::vector< std::vector<int8_t> > &p, std::vector<int> &offsets, int len,
+                             std::vector<int8_t> &q, int offset, int x) {
+    int y = 1, iIdx, oIdx;
+    for (int i = 0; i < (int)p.size(); i++) {
+        std::vector<int8_t> * pi = &p[i];
+        for (iIdx = offsets[i], oIdx = offset; iIdx < offsets[i] + len; iIdx++, oIdx++) {
+            int pij = pi != NULL ? (*pi)[iIdx] & 0x000000FF : 0;
+            q[oIdx] = (int8_t) (q[oIdx] ^ mulTable[pij][y]);
+        }
+        y = mulTable[x][y];
+    }
+}
+
+void GaloisField::substitute(std::vector<ByteBuffer> &p, int len, ByteBuffer &q, int x) {
+    int y = 1, iIdx, oIdx;
+    for (int i = 0; i < (int)p.size(); i++) {
+        ByteBuffer* pi = &p[i];
+        int pos = pi != NULL ? pi->position() : 0;
+        int limit = pi != NULL ? pi->limit() : len;
+        for (oIdx = q.position(), iIdx = pos; iIdx < limit; iIdx++, oIdx++) {
+            int pij = pi != NULL ? pi->get(iIdx) & 0x000000FF : 0;
+            q.put((int8_t)(q.get(oIdx) ^ mulTable[pij][y]), oIdx);
+        }
+        y = mulTable[x][y];
+    }
+}
+
+/**
+ * Substitute x into polynomial p(x).
+ *
+ * @param p input polynomial
+ * @param x input field
+ * @return p(x)
+ */
+int GaloisField::substitute(std::vector<int> &p, int x) {
+    int result = 0;
+    int y = 1;
+    for (int i = 0; i < (int)p.size(); i++) {
+        result = result ^ mulTable[p[i]][y];
+        y = mulTable[x][y];
+    }
+    return result;
+}
+
```

Comment:
If this was called heavily, should do a profile and improve this method's performance